visivo/query/runner.py

```diff
@@ -60,73 +64,71 @@ def run(self):
+            self.failed_job_results.append(job_result)
+
+    def update_job_queue(self, job_tracker: JobTracker):
+        terminal_nodes = [
+            n for n in self.job_dag.nodes() if self.job_dag.out_degree(n) == 0
+        ]
 
-            if dependencies_completed and not job_tracker.is_job_name_enqueued(
-                job.name
+        failed_items = [result.item for result in self.failed_job_results]
+        successful_items = [result.item for result in self.successful_job_results]
+        for terminal_node in terminal_nodes:
+            descendants = all_descendants(dag=self.project_dag, from_node=terminal_node)
+            if terminal_node in successful_items or terminal_node in failed_items:
+                self.job_dag.remove_node(terminal_node)
+                continue
+            elif any(descendant in failed_items for descendant in descendants):
+                Logger.instance().info(
+                    f"Skipping job for {terminal_node} because it has a failed dependency"
+                )
+                self.job_dag.remove_node(terminal_node)
+                continue
+            elif self.name_filter and not family_tree_contains_named_node(
+                item=terminal_node, name=self.name_filter, dag=self.project_dag
             ):
+                self.job_dag.remove_node(terminal_node)
+                continue
+
+            job = self.create_jobs_from_item(terminal_node)
+            if not job:
+                self.job_dag.remove_node(terminal_node)
+            elif not job_tracker.is_job_name_enqueued(job.name):
                 if not job.output_changed and self.run_only_changed:
                     job.future = CachedFuture()
                 job_tracker.track_job(job)
+            else:
+                pass
 
-        return all_dependencies_completed and job_tracker.empty()
-
-    def job_callback(self, future: Future):
-        job_result: JobResult = future.result(timeout=1)
-        if job_result.success:
-            Logger.instance().success(str(job_result.message))
-        else:
-            Logger.instance().error(str(job_result.message))
-            self.errors.append(str(job_result.message))
+        return len(self.job_dag.nodes()) == 0
 
-    def _all_jobs(self) -> List[Job]:
-        jobs = []
-        jobs = jobs + run_trace_jobs(
-            dag=self.dag,
-            output_dir=self.output_dir,
-            project=self.project,
-            name_filter=self.name_filter,
-        )
-        jobs = jobs + csv_script_jobs(
-            dag=self.dag,
-            output_dir=self.output_dir,
-            project=self.project,
-            name_filter=self.name_filter,
-        )
-        jobs = jobs + run_local_merge_jobs(
-            dag=self.dag,
-            output_dir=self.output_dir,
-            project=self.project,
-            name_filter=self.name_filter,
-        )
-        return jobs
+    def create_jobs_from_item(self, item: ParentModel):
```

Comment:
We should add tests to the DAG as well. I dont know if that makes sense to do here or somewhere else.

Reply:
The other thing we should consider adding to the dag is breaking out creating the engine pool. Would enable us to give the user better logs and understand where our run bottlenecks are.

Reply:
I didn't follow the comment around the pools.

Reply:
> We should add tests to the DAG as well. I dont know if that makes sense to do here or somewhere else.

This is a great idea, and I think that isn't a huge lift, but this one is getting long in the tooth.  I think most of the work there is how do we want to trigger the test run since it is part of the larger dag.  Do they run by default?  Do you have to explicitly say not to run them?